### PR TITLE
Add pipeline shader module error enumeration.

### PR DESCRIPTION
**Pull request by jdashg**
Related to #645 shader-maps.

***
<a href="https://pr-preview.s3.amazonaws.com/jdashg/gpuweb/pull/646.html" title="Last updated on May 18, 2020, 9:34 PM UTC (ad1ea78)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/646/5048a91...jdashg:ad1ea78.html" title="Last updated on May 18, 2020, 9:34 PM UTC (ad1ea78)">Diff</a>

### CODE REVIEW

**Comment by kainino0x** [comment]
I'm pretty sure we can have the parser errors in createShaderModule. But what about more complex errors? What if a non-parser error *could* be detected in createShaderModule but it's not immediately obvious whether it should be there or in createPipeline? Or what if an error is not 100% clear whether it's during parsing or later?

For testing I think it's important that it's reliable/portable (unlike in GL).

File: spec/index.bs [hunk 1]
```diff
@@ -2028,13 +2028,48 @@ dictionary GPUProgrammableStageDescriptor {
...
+};
+enum GPUProgrammableStageType {
+    'vertex',
+    'fragment',
+    'compute'
+};
+
+[Serializable]
+interface GPUCompilationMessage {
+    GPUCompilationMessageType type;
```
Comment (litherum) [inline spec/index.bs]: I’m pretty surprised that all these fields are present. Do we expect applications to use these fields at runtime to patch up their shaders dynamically?

**Comment by kvark** [comment]
@kainino0x 
We should at least attempt to draw the line between what errors are possible at what stage. If that becomes blurry, we'll have to make error reporting less rigid.

In my understanding, *most* of the errors needs to be detected at the `createShaderModule` stage:
  - parsing
  - type checking
  - control flow uniformity analysis

The result - the shader module - is basically a set of ready to use entry points with defined interfaces, i.e. all the input vertex attributes, output colors, used storage and uniform buffers, etc. The only validation errors left to `createRenderPipeline` step are about the interface mismatches, as outlined in https://github.com/gpuweb/gpuweb/issues/644#issuecomment-604780355

File: spec/index.bs (hunk 1 above)
Comment (kainino0x) [reply spec/index.bs]: This looks like a structured representation of a standard error message to me. How would you see them being used for patching?

File: spec/index.bs [hunk 2]
```diff
@@ -2028,13 +2028,48 @@ dictionary GPUProgrammableStageDescriptor {
...
+    GPUProgrammableStageType stage;
+    unsigned long long lineNum;
+    unsigned long long linePos;
+    DOMString? mappedFileName;
+    unsigned long long mappedLineNum;
+    unsigned long long mappedLinePos;
+};
```
Comment (damyanp) [inline spec/index.bs:2052]: Is there a message string as well?  I'm probably missing something, but I can't see anything that says what the actual error or warning might be.

**Comment by jdashg** [comment]
From meeting:
* Should spec what happens when compile errors *and* other pipeline creation errors

File: spec/index.bs (hunk 2 above)
Comment (jdashg) [reply spec/index.bs:2052]: Haha, oops!

**Comment by Kangz** [comment]
Reading back the minutes from yesterday, I'm surprised we came to the conclusion that compilation errors should be on the pipeline object and not the shader module. Talking with @dneto0 and @dj2 it seems that all the WGSL compilation errors should be caught at `GPUShaderModule` creation.

Errors left at pipeline creation are mismatches between the WGSL interface and the interface declared in the descriptor. Or backend compilation errors, but I don't know that we should expose them to the user because it seems more like an issue in the browser's translator. Interface mismatch errors are not actionable in the same way as WGSL compilation errors and could be surfaced through push/popErrorScope.

So instead, what do you think of having the following?

```webidl
partial interface GPUShaderModule {
    [SameObject] Promise<sequence<GPUCompilationMessage>> compilationMessages();
};
```

**Comment by kainino0x** [comment]
From that discussion this morning I've decided that I support having errors on ShaderModule. In particular, we would provide structured errors for ShaderModule creation and non-structured (for now) errors for pipeline creation. Reducing that functionality duplication makes it more appealing to me. Pipeline creation errors would be limited as much as possible to compatibility issues (layout compatibility, interface between stages, etc.)

David also shared some feedback from the Vulkan ecosystem saying that this choice in Vulkan has been annoying for users, which is a useful signal.

File: spec/index.bs [hunk 3]
```diff
@@ -2003,6 +2022,11 @@ shader module object, and {{Serializable}} means that the reference can be
 it concurrently. Since {{GPUShaderModule}} is immutable, there are no race
 conditions.
 
+If no source-map was provided for the |GPUShaderModule|:
```
Comment (kvark) [inline spec/index.bs:2025]: this should follow-up after #645 , if it's not about to land

**Comment by Kangz** [comment]
Is anything missing (apart from the removal of the mention of source maps) before merging this?

File: spec/index.bs [hunk 4]
```diff
@@ -1991,8 +1991,28 @@ Shader Modules {#shader-modules}
...
+};
+
+[Serializable]
+interface GPUCompilationInfo {
+    sequence<GPUCompilationMessage> messages;
+};
+
 [Serializable]
 interface GPUShaderModule {
+    [SameObject] Promise<GPUCompilationInfo> compilationInfo();
```
Comment (kainino0x) [inline spec/index.bs:2015]: > The [SameObject] extended attribute must not be used on anything other than a read only attribute whose type is an interface type or object.

Comment (kainino0x) [inline spec/index.bs:2015]:
```suggestion
    Promise<GPUCompilationInfo> compilationInfo();
```

File: spec/index.bs [hunk 5]
```diff
@@ -2003,6 +2023,11 @@ shader module object, and {{Serializable}} means that the reference can be
 it concurrently. Since {{GPUShaderModule}} is immutable, there are no race
 conditions.
 
+If no source-map was provided for the |GPUShaderModule|:
+- {{GPUCompilationMessage/mappedFileName}} will be null
+- {{GPUCompilationMessage/mappedLineNum}} will be 0
+- {{GPUCompilationMessage/mappedLinePos}} will be 0
+
```
Comment (kainino0x) [inline spec/index.bs:2030]: Also this shouldn't be here anymore.

```suggestion
```

**Comment by kainino0x** [comment]
Fixed up some uninteresting issues, and merging.

**Pull request merged**